Cache map option for performance in scrolling use cases

ivanfoong opened the pull request:
I used this component in my app, which includes a map view for every listview item. Had problem with the map view staggering/lagging while scrolling on the listview, so this PR added cache as image option for map view.

lelandrichardson commented on components/MapView.js:
```diff
@@ -77,6 +77,13 @@ var MapView = React.createClass({
+     * Default value is `false`.
+     *
+     */
+    cacheEnabled: PropTypes.bool,
```
Can you add that this is android only?

On that topic... is there anything analogous to this on mapkit that you know of?

ivanfoong replied:
I see a potential solution for iOS as mentioned in http://stackoverflow.com/a/16186766. But currently I'm only using it in Android. Maybe sometime later when I have the time i'll work on this for iOS.

lelandrichardson commented on android/src/main/java/com/AirMaps/AirMapView.java:
```diff
@@ -100,6 +110,20 @@ public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float d
         eventDispatcher = context.getNativeModule(UIManagerModule.class).getEventDispatcher();
+
+        this.mapLoadingTextView = new TextView(context);
+        this.mapLoadingTextView.setText("Loading map...");
```
Is this standard?  What if people are building an app for non-english speakers?  Is there a platform loading indicator that we can use instead?

ivanfoong replied:
Will change this to use loading indicator instead

lelandrichardson commented on android/src/main/java/com/AirMaps/AirMapView.java:
```diff
@@ -100,6 +110,20 @@ public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float d
+
+        this.mapLoadingTextView = new TextView(context);
+        this.mapLoadingTextView.setText("Loading map...");
+        this.mapLoadingTextView.setBackgroundColor(Color.parseColor("#eeeeee"));
```
also ideally this color isn't hard coded... maybe we expose it as a prop with this as a default?

ivanfoong replied:
Will use props to change loading state background color

lelandrichardson commented:
This looks awesome and useful... the caching is a feature I didn't know about.

Can you upload a gif of what this looks like?

ivanfoong commented:
Sure, I'll update so it's more configurable. At the moment I'm only using
it for android, not sure if map snapshot is available on iOS sdk

On Friday, 26 February 2016, Leland Richardson <notifications@github.com>
wrote:

> This looks awesome and useful... the caching is a feature I didn't know
> about.
>
> Can you upload a gif of what this looks like?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lelandrichardson/react-native-maps/pull/91#issuecomment-188939900>
> .
>


-- 
Regards,
Ivan <ivan@d--buzz.com>

ivanfoong commented:
![trafficwatch sg map cache](https://cloud.githubusercontent.com/assets/798022/13386900/66c1a278-deec-11e5-83c5-237a2f2e9dfd.gif)

Sample usage
```
<MapView
          style={{flex: 1, backgroundColor: '#eeeeee'}}
          showsUserLocation={true}
          region={region}
          cacheEnabled={true}
          cacheLoadingBackgroundColor={'#eeeeee'}
          cacheLoadingIndicatorColor={'#666666'}
          showsTraffic={true}>
</MapView>
```

ivanfoong commented:
@lelandrichardson see if this PR checks out? I'll do another PR when I get around to do map caching for iOS

jrichardlai commented:
@ivanfoong thanks for this PR this helped to remove the black blinking and the black lag.
I actually needed to do a modification for my goal ( I still want the loader, to avoid the black screen, but no caching), I renamed `cacheLoadingBackgroundColor` and `cacheLoadingIndicatorColor` to `loadingBackgroundColor` and `loadingColorIndicatorColor` and added a method `loadingEnabled`

jrichardlai commented:
Here are my changes https://github.com/ivanfoong/react-native-maps/compare/master...jrichardlai:android-map-loading-cache?expand=1

ivanfoong commented:
@jrichardlai cool, you mind updating the doc for the prop name change too?

jrichardlai commented:
@ivanfoong are you referring to this doc https://github.com/lelandrichardson/react-native-maps/blob/master/docs/mapview.md ? I didn't see any changes in the PR.
Also do you want to merge/cherry-pick my changes in yours?
Seems the branch needs a PR, I can help you with that if you want.

ivanfoong commented:
Yeah, sorry, I have not gotten around to document the cache functionality in the readme. 

I'll merge your changes to my fork and I should have the time today to fix my PR on the conflict issue.

jrichardlai commented:
@ivanfoong thanks for merging my changes, it looks good, I'm gonna test it more today.

jrichardlai commented:
@ivanfoong looks good to me! Good job! I tested on multiple devices works well.
Realized though that my initial problem of black blinking still occurs on the first load of a map in the app, but this is a different issue.

I added examples https://github.com/ivanfoong/react-native-maps/compare/master...jrichardlai:add-cache-loading-example?expand=1

ivanfoong commented:
@jrichardlai :+1: Nice! Thanks for the contribution! I kinda digressed on creating the examples. 

Do you happen to also do iOS? I guess @lelandrichardson is sitting on merging this PR as these features only works for Android.

jrichardlai commented:
@ivanfoong not really, my experience with iOS and Android are mostly from hacking since React Native.

ivanfoong commented:
@jrichardlai icic, no problem. I'll try to put in sometime this week to try getting it up on iOS too.

ivanfoong commented:
Hi @lelandrichardson, I managed get map caching and loading indicator done for iOS, let me know if there's any changes needed. cc @jrichardlai

lelandrichardson commented on ios/AirMaps/AIRMap.m:
```diff
@@ -67,6 +73,23 @@ - (instancetype)init
         self.calloutView = [SMCalloutView platformCalloutView];
         self.calloutView.delegate = self;
+        
+        self.cacheImageView = [[UIImageView alloc] initWithFrame:self.bounds];
```
Is instantiating all of these views (imageview + view + indicatorview) potentially a bit wasteful if these props aren't there?

ivanfoong replied:
I'll see if i can refactor these views for lazy loading

lelandrichardson commented on ios/AirMaps/AIRMap.m:
```diff
@@ -67,6 +73,23 @@ - (instancetype)init
+        [self.loadingView addSubview:self.activityIndicatorView];
+        self.activityIndicatorView.center = self.loadingView.center;
+        self.activityIndicatorView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
+        [self.activityIndicatorView startAnimating];
```
if the loadingEnabled prop is false by default, do we want to start the animation in every case? or make the view visible?  

I'm thinking for this feature we would want to add this capability without changing the previous behavior so that it's an enhancement rather than a "breaking" change / difference.

lelandrichardson commented:
wow @ivanfoong this is excellent work! Thanks for doing this. I had a couple of lines of feedback. Have any thoughts?

Based on your gif, this looks like it's probably good to merge in, but I'm just trying to be conscious of adding more weight than is needed (though admittedly probably a microoptimization...).

lelandrichardson commented:
cc @jrichardlai @christopherdro

ivanfoong commented:
@lelandrichardson I'll work on refactoring the views to lazy loading, should also address the concerns you have and improve the performance

ivanfoong commented:
@lelandrichardson I have done minor changes to make imageview, view, indicatorview lazy load, they will only be created if `loadingEnabled`/`cachingEnabled` props has been set to true at least once.

ivanfoong commented:
I have just added the codes to allow updating of the image cached when the mapview size changes (e.g. device orientation changes)

christopherdro commented:
LGTM!

@ivanfoong How have you been testing this?

ivanfoong commented:
@christopherdro I actually added additional codes to invert the colors of the cached image to prove it's working, but i'm planning to change the example to show as a list view of mapviews to demo the scrolling performance issue, with and without the cache option

ivanfoong commented:
Just updated cache map example with MapView of countries in a ListView

jrichardlai commented:
@ivanfoong I looked at the cache example its pretty cool! However it seems that the map loading was broken for android I have a fix https://github.com/ivanfoong/react-native-maps/compare/master...jrichardlai:fix-map-loading-android?expand=1

ivanfoong commented:
Oh yeah, I missed that 1 out.

jrichardlai commented on android/src/main/java/com/AirMaps/AirMapView.java:
```diff
@@ -256,6 +293,65 @@ public void setShowsUserLocation(boolean showUserLocation) {
+                this.mapLoadingProgressBar.setSecondaryProgressTintList(secondaryProgressTintList);
+                this.mapLoadingProgressBar.setIndeterminateTintList(indeterminateTintList);
+            } else {
+                int color = ThemeUtils.getThemeAttrColor(getContext(), android.R.attr.color);
```
@ivanfoong I was updating my branch using the new changes you made and it seem that I get an error here `Caused by: java.lang.NoClassDefFoundError: android.support.v7.internal.widget.ThemeUtils` what is required to use ThemeUtils?

ivanfoong replied:
@lelandrichardson I was trying to use the default color of progressBar for android. Would including additional dependency on Android Support Library V7 to use as default color be better, or to hard code the color to a fixed default be better?

jrichardlai replied:
I think it's better to not add more dependencies, this could cause some issues if other devs use in their project a different version of android support

ivanfoong replied:
Agreed, I'm updating it to use a default color instead.

jrichardlai commented:
@ivanfoong also I did some tests and it seems that the press on the cached map on iOS is not propagated.

This used to work:
```
<TouchableWithoutFeedback onPress={this.showMap}>
  <MapView
    style={[styles.map]}
    scrollEnabled={false}
    zoomEnabled={false}
    cacheEnabled={true}
    locationDataObjects={this.locationDataObjects()}
  />
</TouchableWithoutFeedback>
```
But had to do this to fix it.
```
<TouchableWithoutFeedback onPress={this.showMap}>
  <View>
    <MapView
      style={[styles.map]}
      scrollEnabled={false}
      zoomEnabled={false}
      cacheEnabled={true}
      locationDataObjects={this.locationDataObjects()}
    />
  </View>
</TouchableWithoutFeedback>
```

ivanfoong commented:
You're right! I have set `userInteractionEnabled=No` when using caching, was trying to get MapView not to scroll when it's in a ScrollView (to scroll ScrollView while dragging on the MapView). You have any suggestion on how to implement this to fit both use case?

ivanfoong commented:
@lelandrichardson just merged with the latest master, but i'm having issue with the new react and react-native library, lots of naming collision between react and react-native library, i'll dive deeper and report on this

> Failed to build DependencyGraph: @providesModule naming collision:
  Duplicate module name: ReactNativeTextComponent

lelandrichardson commented:
@ivanfoong this is likely from having multiple versions of react.  run this in your console:
```
npm dedupe react-native
npm dedupe react
```

or, alternatively, just run `rm -rf node_modules && npm install`

ivanfoong commented:
hmm, tried `npm dedupe` and also `rm -rf node_modules && npm install` in `example` folder, but both still results in the following for me.

> Failed to build DependencyGraph: @providesModule naming collision:
  Duplicate module name: TouchHistoryMath
  Paths: /Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/react/lib/TouchHistoryMath.js collides with /Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/react-native/Libraries/vendor/react/browser/eventPlugins/TouchHistoryMath.js
This error is caused by a @providesModule declaration with the same name accross two different files.

>Error: @providesModule naming collision:
  Duplicate module name: TouchHistoryMath
  Paths: /Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/react/lib/TouchHistoryMath.js collides with /Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/react-native/Libraries/vendor/react/browser/eventPlugins/TouchHistoryMath.js

>This error is caused by a @providesModule declaration with the same name accross two different files.
    at HasteMap._updateHasteMap (/Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/node-haste/lib/DependencyGraph/HasteMap.js:160:15)
    at /Users/ivanfoong/Repos/ivanfoong/react-native-maps/example/node_modules/node-haste/lib/DependencyGraph/HasteMap.js:125:25

jrichardlai commented:
@ivanfoong which version of react and react-native are installed? It worked for me when I was on react `0.14.6`

ivanfoong commented:
I'm on `react@15.0.2` and `react-native@0.25.1` as the example's dependency was set to `latest`, i'll try with `0.14.6`

jrichardlai commented:
@ivanfoong woot works good now!

ivanfoong commented:
👍 finally my first RN PR done!

jrichardlai commented:
@ivanfoong I'm not sure if the image is really cached for iOS it seems that is actually still showing the map and not just a cached image ( I modified a little the code and it seems that if i allow scrolling and zooming i am still able to interact with the map )

ivanfoong commented:
@jrichardlai Let me test again and get back to you

ivanfoong commented:
@jrichardlai try the branch at https://github.com/ivanfoong/react-native-maps/tree/test_cache_working
it will show a "negative" cached image instead to show that it's indeed an image instead of the actual mapview

jrichardlai commented:
@ivanfoong thanks for the explanation, it looks good! Awesome job. @lelandrichardson I think this is gtg!

ivanfoong commented:
@jrichardlai yeah, you're right. I have just updated the code to store scrollingEnabled and zoomEnabled configuration and will restore them upon disabling cache. I also updated the `CachedMap` example to allow toggling of `cacheEnabled` prop

jrichardlai commented:
@ivanfoong rebasing would be hard for this PR android structure has been changed, probably squashing them together would be the best. I can do that for you @ivanfoong

ivanfoong commented:
@jrichardlai agreed, please help me do that

ivanfoong commented:
Squashed the changes and opened new pull request #295
